arbitrum-docs/launch-orbit-chain/how-tos/customize-arbos.mdx

```diff
@@ -0,0 +1,141 @@
+arbos.L2ToL1TxEventID = ArbSys.events["L2ToL1Tx"].template.ID
+```
+You need to append the following code to it:
+```go
+ArbSys := insert(MakePrecompile(pgen.ArbSysMetaData, &ArbSys{Address: types.ArbSysAddress}))
+arbos.ArbSysAddress = ArbSys.address
+arbos.L2ToL1TransactionEventID = ArbSys.events["L2ToL1Transaction"].template.ID
+arbos.L2ToL1TxEventID = ArbSys.events["L2ToL1Tx"].template.ID
+// The arbos version control logic
+ArbOwner.methodsByName["SayHi"].arbosVersion = ${The arbos version you want to activate this method}
+```
+In this way, this method will be executed normally and return results only after you update ArbOS to the target version.
+
+### 2. Create a new precompile contract on a specific ArbOS version
+After you add a new precompile named `ArbHi` according to the guide in [customize precompile option 2](./customize-precompile.mdx#option-2-create-a-new-precompile) and make changes to [precompile.go][precompile_main_file_link], 
+you also need to make the following changes:
+```go
+ArbHi := insert(MakePrecompile(pgen.ArbHiMetaData, &ArbHi{Address: types.ArbHiAddress})) // types.ArbHiAddress here is an example address
+// Set activate version to the precompile
+ArbHi.arbosVersion = ${The arbos version you want to activate this precompile}
+// Set activate version to all method
+for _, method := range ArbHi.methods {
+   method.arbosVersion = ${The arbos version you want to activate this precompile}
+}
+```
+In this way, ArbHi and all its methods will be activated after the ArbOS version you set.
+
+### 3. Create a new ArbOS state on a specific ArbOS version
+
+After you add a new state `myNumber` according to the guide in [customize precompile Option 5](./customize-precompile.mdx#option-5-call-and-modify-state), you also need to rewrite [UpgradeArbosVersion][arbosstate_updatearbosversion_method_link] in [arbosstate.go][arbosstate_file_link]:
+Add your expected ArbOS version to the switch case statement of nextArbosVersion. Here we will take ArbOS V21 as an example:
+```go
+ensure := func(err error) {
+    if err != nil {
+        message := fmt.Sprintf(
+            "Failed to upgrade ArbOS version %v to version %v: %v",
+            state.arbosVersion, state.arbosVersion+1, err,
+        )
+        panic(message)
+    }
+}
+nextArbosVersion := state.arbosVersion + 1
+switch nextArbosVersion {
+    case 1:
+        //.....
+    case 2:
+        //....
+    //.....
+    case 21:
+        // Set your new ArbOS state value here
+        ensure(state.SetNewMyNumber(${random number}))
+    //....
+} 
+```
+
+Here, we will ensure that the initial value of $(a random number) after ArbOS is upgraded to V21.
+
+
+:::caution
+It should be noted that when you initialize the state (initial code is in [customize precompile Option 5](./customize-precompile.mdx#option-5-call-and-modify-state)), you need to initialize it to 0 or null value first to avoid potential blockchain reorg.
```

Comment:
```suggestion
It should be noted that when you initialize the state (initial code is in [customize precompile Option 5](./customize-precompile.mdx#option-5-call-and-modify-state)), you need to initialize it to 0 or null value first to avoid a potential re-org on your chain.
```